src/lib.rs

```diff
@@ -121,7 +121,7 @@ macro_rules! impl_elastic_array {
 		impl<T> $name<T> where T: Copy {
 			pub fn new() -> $name<T> {
 				$name {
-					raw: $dummy::Arr(unsafe { $crate::core_::mem::uninitialized() }),
+					raw: $dummy::Arr(unsafe { $crate::core_::mem::MaybeUninit::uninit().assume_init() }),
```

Comment:
From the documentation of `MaybeUnit::assume_init`:

_It is up to the caller to guarantee that the MaybeUninit<T> really is in an initialized state. Calling this when the content is not yet fully initialized causes immediate undefined behavior.

Clearly, we call `assume_init()` on unitialized memory.

Thus,  doesn't seem like the right fix. It would be nice if `LLVM` could optimize use of `zeros` instead as we did in `parity_common`. However, as this is generic of the data in the `array` doesn't seem possible with requiring `T: Default` everywhere but I'm not sure if  `LLVM` will optimize it in that case.

Reply:
Yeah I’m not sure what kind of bounds we’d need here. I peeked at what smallvec does and it seems pretty similar (but this stuff is subtle).
Any suggestion of things I should try?

Reply:
This should do it:

```git
diff --git a/src/lib.rs b/src/lib.rs
index 72efa09..71645b6 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -85,7 +85,7 @@ macro_rules! impl_elastic_array {
 			fn cmp(&self, other: &Self) -> Ordering {
 				(&**self).cmp(&*other)
 			}
-		} 
+		}
 
 		impl<T> Hash for $name<T> where T: Hash {
 			fn hash<H>(&self, state: &mut H) where H: Hasher {
@@ -112,16 +112,16 @@ macro_rules! impl_elastic_array {
 			}
 		}
 
-		impl<T> Default for $name<T> where T: Copy {
+		impl<T: Default> Default for $name<T> where T: Copy {
 			fn default() -> $name<T> {
 				Self::new()
 			}
 		}
 
-		impl<T> $name<T> where T: Copy {
+		impl<T> $name<T> where T: Copy + Default {
 			pub fn new() -> $name<T> {
 				$name {
-					raw: $dummy::Arr(unsafe { $crate::core_::mem::uninitialized() }),
+					raw: $dummy::Arr([T::default();  $size]),
 					len: 0
 				}
 			}
@@ -176,7 +176,7 @@ macro_rules! impl_elastic_array {
 			}
 
 			pub fn clear(&mut self) {
-				self.raw = $dummy::Arr(unsafe { $crate::core_::mem::uninitialized() });
+				self.raw = $dummy::Arr([T::default(); $size]);
 				self.len = 0;
 			}
 
@@ -190,7 +190,7 @@ macro_rules! impl_elastic_array {
 					$dummy::Arr(a) => {
 						let mut vec = Vec::new();
 						vec.reserve(self.len);
-						unsafe {	
+						unsafe {
 							$crate::core_::ptr::copy(a.as_ptr(), vec.as_mut_ptr(), self.len);
 							vec.set_len(self.len);
 						}
@@ -208,7 +208,7 @@ macro_rules! impl_elastic_array {
 				if elen == 0 {
 					return;
 				}
-				
+
 				let len = self.len;
 				assert!(index <= len);
 
@@ -232,14 +232,14 @@ macro_rules! impl_elastic_array {
 							let ob = self.raw.slice().as_ptr();
 							let ep = elements.as_ptr();
 							let oe = ob.offset(index as isize);
-							
+
 							// copy begining of an array
 							ptr::copy(ob, p, index);
 
 							// copy new elements
 							ptr::copy(ep, p.offset(index as isize), elen);
 
-							// copy end of an array	
+							// copy end of an array
 							ptr::copy(oe, p.offset((index + elen) as isize), len - index);
 						}
 						vec.set_len(self.len + elen);
@@ -306,7 +306,7 @@ macro_rules! impl_elastic_array {
 			}
 		}
 
-		impl<'a, T: 'a + Copy> From<&'a [T]> for $name<T> {
+		impl<'a, T: 'a + Copy + Default> From<&'a [T]> for $name<T> {
 			fn from(s: &'a [T]) -> Self { Self::from_slice(s) }
 		}
 	)

``

Reply:
more or less requires `Copy + Default` to be constructed before just `Copy`